kv/dist_sender.go

```diff
@@ -203,67 +203,79 @@ func NewDistSender(ctx *DistSenderContext, gossip *gossip.Gossip) *DistSender {
-				if proto.IsRead(args) && !perm.CanRead(header.User) {
-					return false, nil
+		// Check for admin methods.
+		if proto.IsAdmin(args) {
+			if header.User != security.RootUser {
+				return util.Errorf("user %q cannot invoke admin command %s", header.User, args.Method())
+			}
+			return nil
+		}
+		// Get permissions map from gossip.
+		configMap, err := ds.gossip.GetInfo(gossip.KeyConfigPermission)
+		if err != nil {
+			return util.Errorf("permissions not available via gossip")
+		}
+		if configMap == nil {
+			return util.Errorf("perm configs not available; cannot execute %s", args.Method())
+		}
+		permMap := configMap.(config.PrefixConfigMap)
+		headerEnd := header.EndKey
+		if len(headerEnd) == 0 {
+			headerEnd = header.Key
+		}
+		// Visit PermConfig(s) which apply to the method's key range.
+		//   - For each perm config which the range covers, verify read or writes
+		//     are allowed as method requires.
+		//   - Verify the permissions hierarchically; that is, if permissions aren't
+		//     granted at the longest prefix, try next longest, then next, etc., up
+		//     to and including the default prefix.
+		//
+		// TODO(spencer): it might make sense to visit prefixes from the
+		//   shortest to longest instead for performance. Keep an eye on profiling
+		//   for this code path as permission sets grow large.
+		if err := permMap.VisitPrefixes(header.Key, headerEnd,
+			func(start, end proto.Key, cfg gogoproto.Message) (bool, error) {
+				hasPerm := false
+				if err := permMap.VisitPrefixesHierarchically(start, func(start, end proto.Key, cfg gogoproto.Message) (bool, error) {
+					perm := cfg.(*config.PermConfig)
+					if proto.IsRead(args) && !perm.CanRead(header.User) {
+						return false, nil
+					}
+					if proto.IsWrite(args) && !perm.CanWrite(header.User) {
+						return false, nil
+					}
+					// Return done = true, as permissions have been granted by this config.
+					hasPerm = true
+					return true, nil
+				}); err != nil {
+					return false, err
 				}
-				if proto.IsWrite(args) && !perm.CanWrite(header.User) {
-					return false, nil
+				if !hasPerm {
+					if len(header.EndKey) == 0 {
+						return false, util.Errorf("user %q cannot invoke %s at %q", header.User, args.Method(), start)
+					}
+					return false, util.Errorf("user %q cannot invoke %s at %q-%q", header.User, args.Method(), start, end)
 				}
-				// Return done = true, as permissions have been granted by this config.
-				hasPerm = true
-				return true, nil
+				return false, nil
 			}); err != nil {
-				return false, err
-			}
-			if !hasPerm {
-				if len(header.EndKey) == 0 {
-					return false, util.Errorf("user %q cannot invoke %s at %q", header.User, args.Method(), start)
-				}
-				return false, util.Errorf("user %q cannot invoke %s at %q-%q", header.User, args.Method(), start, end)
-			}
-			return false, nil
-		})
+			return err
+		}
+		return nil
+	}
+
+	for i, l := 0, len(bArgs.Requests); i < l; i++ {
```

Comment:
that works. Even `for i := ....` would be better than this here. Probably had different code in here at some point.